### PR TITLE
chore(operations)!: Teach Environment to build for other architectures.

### LINKED ISSUE #3311: Nightly builds are failing
As seen [here](https://github.com/timberio/vector/runs/934492571?check_suite_focus=true):

```
build-armv7-unknown-linux-musleabihf_1            | error[E0512]: cannot transmute between types of different sizes, or dependently-sized types
build-armv7-unknown-linux-musleabihf_1            |  --> /opt/rust/cargo/registry/src/github.com-1ecc6299db9ec823/cidr-utils-0.4.2/src/cidr/v4/ipv4_cidr_error.rs:4:45
build-armv7-unknown-linux-musleabihf_1            |   |
build-armv7-unknown-linux-musleabihf_1            | 4 | #[derive(Debug, Copy, Clone, Eq, PartialEq, Ordinalize)]
build-armv7-unknown-linux-musleabihf_1            |   |                                             ^^^^^^^^^^
build-armv7-unknown-linux-musleabihf_1            |   |
build-armv7-unknown-linux-musleabihf_1            |   = note: source type: `std::mem::Discriminant<cidr::v4::ipv4_cidr_error::Ipv4CidrError>` (32 bits)
build-armv7-unknown-linux-musleabihf_1            |   = note: target type: `u64` (64 bits)
build-armv7-unknown-linux-musleabihf_1            |   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
build-armv7-unknown-linux-musleabihf_1            | 
build-armv7-unknown-linux-musleabihf_1            | error[E0512]: cannot transmute between types of different sizes, or dependently-sized types
build-armv7-unknown-linux-musleabihf_1            |  --> /opt/rust/cargo/registry/src/github.com-1ecc6299db9ec823/cidr-utils-0.4.2/src/cidr/v6/ipv6_cidr_error.rs:4:45
build-armv7-unknown-linux-musleabihf_1            |   |
build-armv7-unknown-linux-musleabihf_1            | 4 | #[derive(Debug, Copy, Clone, Eq, PartialEq, Ordinalize)]
build-armv7-unknown-linux-musleabihf_1            |   |                                             ^^^^^^^^^^
build-armv7-unknown-linux-musleabihf_1            |   |
build-armv7-unknown-linux-musleabihf_1            |   = note: source type: `std::mem::Discriminant<cidr::v6::ipv6_cidr_error::Ipv6CidrError>` (32 bits)
build-armv7-unknown-linux-musleabihf_1            |   = note: target type: `u64` (64 bits)
build-armv7-unknown-linux-musleabihf_1            |   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
build-armv7-unknown-linux-musleabihf_1            | 
build-armv7-unknown-linux-musleabihf_1            | error[E0512]: cannot transmute between types of different sizes, or dependently-sized types
build-armv7-unknown-linux-musleabihf_1            |  --> /opt/rust/cargo/registry/src/github.com-1ecc6299db9ec823/cidr-utils-0.4.2/src/cidr/ip_cidr_error.rs:6:45
build-armv7-unknown-linux-musleabihf_1            |   |
build-armv7-unknown-linux-musleabihf_1            | 6 | #[derive(Debug, Copy, Clone, Eq, PartialEq, Ordinalize)]
build-armv7-unknown-linux-musleabihf_1            |   |                                             ^^^^^^^^^^
build-armv7-unknown-linux-musleabihf_1            |   |
build-armv7-unknown-linux-musleabihf_1            |   = note: source type: `std::mem::Discriminant<cidr::ip_cidr_error::IpCidrError>` (32 bits)
build-armv7-unknown-linux-musleabihf_1            |   = note: target type: `u64` (64 bits)
build-armv7-unknown-linux-musleabihf_1            |   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
build-armv7-unknown-linux-musleabihf_1            | 
build-armv7-unknown-linux-musleabihf_1            | error: aborting due to 3 previous errors
build-armv7-unknown-linux-musleabihf_1            | 
build-armv7-unknown-linux-musleabihf_1            | For more information about this error, try `rustc --explain E0512`.
build-armv7-unknown-linux-musleabihf_1            | error: could not compile `cidr-utils`.
build-armv7-unknown-linux-musleabihf_1            | 
build-armv7-unknown-linux-musleabihf_1            | To learn more, run the command again with --verbose.
build-armv7-unknown-linux-musleabihf_1            | warning: build failed, waiting for other jobs to finish...
build-armv7-unknown-linux-musleabihf_1            | error: build failed
make: *** [build-armv7-unknown-linux-musleabihf] Error 101
```

### PR DESCRIPTION
**Pull request by Hoverbear**
Closes #3311

This introduces a simplified cross-compiler toolchain for us.

After removing our Armv7 builds in #3352 we could opt for a much, much simpler option with https://github.com/richfelker/musl-cross-make. This introduces the components to our environment, as well as adds tooling sufficient for our contributors to use it locally on most Linux machines.

Please note this process is a tad rough around the edges still, I plan to touch it up over the coming months. :)

As a consequence of #2824 we needed to teach our Environment to be multi-distro. I added this as well, it currently acts as CentOS7 during gnu builds, and Ubuntu 20.04 for MUSL builds.

## Notable Changes

* `make build-{x86_64, aarch64}-*` are altered and now build outside of docker containers by default. They use musl-cross-make and you can use `ENVIRONMENT=true` to run them in our friendly environment.
* `make build-armv7-*` is gone as per #3352
* I cleaned up `make check-examples` since it was failing.
* You can now specify `ENVIRONMENT_BASEIMAGE` with a docker container (default `ubuntu:20.04`). The environment will try to build from that baseimage by searching the `scripts/environments/bootstrap/` folder. You can also override with `ENVIRONMENT_BOOTSTRAP` set to a shell script.
* This increases our podman support. I think these should work!
* The new cross toolchains default to `scripts/environment/cross` which is part of our `.gitignore`. You can set a different place with `MUSL_CROSS_MAKE` environment variable. I personally cloned musl-cross-make into my normal Git trees and set this environment variable.
* CI jobs are added for build passes of this new process. This will be expanded later to include things like testing. I expect all our `test` and `check` jobs will collapse down into these new jobs in the future.
* A `RELEASE` flag defaulting to `true` is available. Use `RELEASE=false` to dramatically speed up `build-*` jobs.

## Important notes

* We originally talked about storing the toolchains in `target` but this lead to a poor user experience. Crosschains take up to 15 minutes to build so it's better to keep them outside of this folder we frequently wipe.
* We have long term goals to replace our `nightly` and `release` workflows with per-commit/PR release flows that are published when we tag. This is pulling us in that direction!

## Remaining burndown

* [ ] Docs
* [ ] Ensure CI is green
* [ ] Explore testing builds.
* [ ] Clean up UX. (Warn if doing a cross  build without the toolchain, warn if doing builds not on linux)

### CODE REVIEW

**Comment by Hoverbear** [comment]
Giving this a test in our docker environment

File: Makefile [hunk 1]
```diff
@@ -159,19 +243,64 @@ build: ## Build the project in release mode (Supports `ENVIRONMENT=true`)
...
-build-armv7-unknown-linux-musleabihf: load-qemu-binfmt ## Build static binary in release mode for the armv7 architecture
-	$(RUN) build-armv7-unknown-linux-musleabihf
-
-build-aarch64-unknown-linux-musl: load-qemu-binfmt ## Build static binary in release mode for the aarch64 architecture
-	$(RUN) build-aarch64-unknown-linux-musl
+build-all: build-x86_64-unknown-linux-gnu build-x86_64-unknown-linux-musl build-armv7-unknown-linux-musleabihf build-aarch64-unknown-linux-musl ## Build the project in release mode for all supported platforms
+
+build-x86_64-unknown-linux-gnu: ${VECTOR_BINARY_x86_64_unknown_linux_gnu} ## Build dynamically linked binary in release mode for the x86_64 architecture
+
+${VECTOR_BINARY_x86_64_unknown_linux_gnu}:
```
Comment (Hoverbear) [inline Makefile:257]: Noting I need to teach the environment to use centos for this one.

**Comment by Hoverbear** [comment]
I'm pulling in the ideas of https://github.com/timberio/vector/pull/3257/files here

**Comment by Hoverbear** [comment]
Also pulled in #3352

**Comment by Hoverbear** [comment]
I did some feature flag tweaking so I'm running a full build suite locally as well as up here on CI. :)

**Comment by Hoverbear** [comment]
rbenv is struggling because the ubuntu distro has homebrew installed...

**Comment by Hoverbear** [comment]
This is now working for me locally!

**Comment by binarylogic** [comment]
@Hoverbear can you link this to the issues it closes?

**Comment by Hoverbear** [comment]
@binarylogic This would close #2971 and part of #3261 

It'd also accomplish #2899's original goal

**Comment by Hoverbear** [comment]
So this **mostly** works.

All the makefile integration and the vast majority of the building works.

As far as I can tell, **the problem lies in the Glibc to Musl linking.**

So we need to support the following goals:

* Centos:7 base to glibc x86 build.
* Anything to musl x86 build. (We use Ubuntu and musl-cross-make here)
* Anything to musl aarch64 build. (We use Ubuntu and musl-cross-make here)

Further goals are out of scope at this time, but this style of approach allows for our environment to learn to cross compile to other architectures in the future. 

We also looked at using `docker buildx` in #3257 but it conflicts with our goal in #781 .

So when you build the musl builds, there's a problem that (I think) `sasl2` or `rdkafka`'s linkage where it's getting built with the wrong libc.

I am fairly certain that this is a result of us doing something wrong with the C++ related cross compiling variables, as I am not very experienced here. As far as I know, all the Rust and C, and the **majority** of the C++, is working right, but it seems one thing is not.

**Comment by Hoverbear** [comment]
We're opting for #3657

**Pull request closed without merging**